zetaclient/chains/ton/signer/signer.go

```diff
@@ -0,0 +1,241 @@
+}
+
+// Signer represents TON signer.
+type Signer struct {
+	*base.Signer
+	client          LiteClient
+	gateway         *toncontracts.Gateway
+	signaturesCache *lru.Cache
+}
+
+// Signable represents a message that can be signed.
+type Signable interface {
+	Hash() ([32]byte, error)
+	SetSignature(sig [65]byte)
+}
+
+// Outcome possible outbound processing outcomes.
+type Outcome string
+
+const (
+	Invalid Outcome = "invalid"
+	Fail    Outcome = "fail"
+	Success Outcome = "success"
+)
+
+const signaturesHashSize = 1024
+
+var _ interfaces.ChainSigner = (*Signer)(nil)
+
+// New Signer constructor.
+func New(baseSigner *base.Signer, client LiteClient, gateway *toncontracts.Gateway) *Signer {
+	sigCache, _ := lru.New(signaturesHashSize)
+
+	return &Signer{
+		Signer:          baseSigner,
+		client:          client,
+		gateway:         gateway,
+		signaturesCache: sigCache,
+	}
+}
+
+// TryProcessOutbound tries to process outbound cctx.
+// Note that this API signature will be refactored in orchestrator V2
+func (s *Signer) TryProcessOutbound(
+	ctx context.Context,
+	cctx *cc.CrossChainTx,
+	proc *outboundprocessor.Processor,
+	outboundID string,
+	_ interfaces.ChainObserver,
+	zetacore interfaces.ZetacoreClient,
+	zetaBlockHeight uint64,
+) {
+	proc.StartTryProcess(outboundID)
+
+	defer func() {
+		proc.EndTryProcess(outboundID)
+	}()
+
+	outcome, err := s.ProcessOutbound(ctx, cctx, zetacore, zetaBlockHeight)
+	if err != nil {
+		s.Logger().Std.Error().
+			Err(err).
+			Str("outbound.id", outboundID).
+			Uint64("outbound.nonce", cctx.GetCurrentOutboundParam().TssNonce).
+			Str("outbound.outcome", string(outcome)).
+			Msg("Unable to ProcessOutbound")
+	}
+}
+
+// ProcessOutbound signs and broadcasts an outbound cross-chain transaction.
+func (s *Signer) ProcessOutbound(
+	ctx context.Context,
+	cctx *cc.CrossChainTx,
+	zetacore interfaces.ZetacoreClient,
+	zetaHeight uint64,
+) (Outcome, error) {
+	// TODO: note that *InboundParams* are use used on purpose due to legacy reasons.
+	// https://github.com/zeta-chain/node/issues/1949
+	if cctx.InboundParams.CoinType != coin.CoinType_Gas {
+		return Invalid, errors.New("only gas coin outbounds are supported")
+	}
+
+	params := cctx.GetCurrentOutboundParam()
+
+	// TODO: add compliance check
+	// https://github.com/zeta-chain/node/issues/2916
+
+	receiver, err := ton.ParseAccountID(params.Receiver)
+	if err != nil {
+		return Invalid, errors.Wrapf(err, "unable to parse recipient %q", params.Receiver)
+	}
+
+	withdrawal := &toncontracts.Withdrawal{
+		Recipient: receiver,
+		Amount:    params.Amount,
+		// #nosec G115 always in range
+		Seqno: uint32(params.TssNonce),
+	}
+
+	lf := map[string]any{
+		"outbound.recipient": withdrawal.Recipient.ToRaw(),
+		"outbound.amount":    withdrawal.Amount.Uint64(),
+		"outbound.nonce":     withdrawal.Seqno,
+	}
+
+	s.Logger().Std.Info().Fields(lf).Msg("Signing withdrawal")
+
+	if err = s.SignMessage(ctx, withdrawal, zetaHeight, params.TssNonce); err != nil {
```

Comment:
To confirm: there will be no extra relayer address in TON, the TSS itself will be the signer who signs the withdraw tx and also pays the gas fee?

Reply:
That sounds cool!! It means the `gateway` smart contract itself pays gas fee for a call on its `withdraw` method.